README.rst
zps - A zope processes inspector
================================

**zps** is addressed to the manager of zope and plone sites that want to get quickly
get informations about the running instances of zope.

Installation
------------
::

  easy_install psutil
  easy_install rt.zps

Thanks to `@der_sven_ <http://twitter.com/der_sven_>`_ you can find up to date **debian packages** here:

- http://www.apt.leftxs.org/

Usage
-----
**zps** basic usage is straightforward: ::

  [user@localhost]$zps
  CWD:      /home/user/plone1
  User:     user
  PID:      1234
  Conf:     /home/user/plone1/parts/instance/etc/zope.conf
  Address:  8081
  Memory:   1.96%

  CWD:      /home/user/plone2
  User:     user
  PID:      12345
  Conf:     /home/user/plone2/parts/instance/etc/zope.conf
  Address:  8082
  Memory:   1.96%

The default action is to report, for each running instance of zope that it founds:
 
* the Current Working Directory (CWD)
* the id of the user who is running the process
* the process PID
* the zope configuration file used by the instance
* the port (as calculated parsing the configuration file)
* the memory usage

To obtain usage information for **zps** just type zps --help, you will get this::

  [user@localhost]$zps
  zps - report a snapshot of the current zope processes.

  USAGE:
  zps
  zps --help
  zps --pid 1111
  zps --port 8080
  zps --grep plone

As you can see **zps** can filter the results per **zps** or port, e.g., if calling **zps**
returns the report about the two instances above, you will have the following::

  [user@localhost]$zps --pid 1234
  CWD:      /home/user/plone1
  User:     user
  PID:      1234
  Conf:     /home/user/plone1/parts/instance/etc/zope.conf
  Address:  8081
  Memory:   1.96%

  [user@localhost]$zps --port 8082
  CWD:      /home/user/plone2
  User:     user
  PID:      12345
  Conf:     /home/user/plone2/parts/instance/etc/zope.conf
  Address:  8082
  Memory:   1.96%

In the case no instance is found, you will have::

  [user@localhost]$zps
  No running zope instance found

**TODO**: filter output information
It may happen that your server starts to be crowded. In that case the output of 
**zps** is quickly parsable.
Two command line flags should then be introduced:
 
* --show
* --hide

**TODO**: colorize output
It would be a nice feature to highlight with colors the memory usage. 
If the output is in the range:

* 0%-5%: green
* 0%-10%: yellow
* 10%-20%: red
* 20%-100%: blinking red

Examples
--------
::

  [user@localhost]$zps --hide CWD,User,Memory
  PID:      1234
  Conf:     /home/user/plone1/parts/instance/etc/zope.conf
  Address:  8081

  PID:      12345
  Conf:     /home/user/plone2/parts/instance/etc/zope.conf
  Address:  8082

  [user@localhost]$zps --hide CWD,User,Memory
  CWD:      /home/user/plone1
  User:     user
  Memory:   1.96%

  CWD:      /home/user/plone2
  User:     user
  Memory:   1.96%



docs/HISTORY.rst
Changelog
=========

1.0.0rc2 (unreleased)
---------------------

- Nothing changed yet.


1.0.0rc1 (2015-11-13)
---------------------

* Support recent psutil versions
  [ale-rt]

0.9.6 (2014-03-20)
------------------

* psutils 2 compatibility [ale-rt]

0.9.5 (2014-02-14)
------------------

* Python3 compatibility [ale-rt]

0.9.4 (2012-10-29)
------------------

* fixing a crash when zope.conf file is not found [ale-rt]
* handling help flags, fixes #1, fixes #2 [ale-rt]
* cosmetic fixes [ale-rt]

0.9.3 (2012/02/10)
------------------

* moved to github [ale-rt]
* heavy refactoring [ale-rt]
* added --grep flag [ale-rt]
* fix address calculations [ale-rt]
* sorting process list [ale-rt]
* tests [ale-rt]

0.9.2 (2010/06/10)
------------------

* Fixed documentation [keul]
* Fixed computing addresses in the form of ip:port [ale-rt]

0.9.1 (2010/06/10)
------------------

* Fixed egg structure [fdelia]
* pid option [ale-rt, fdelia]
* port option [ale-rt, fdelia]
* Refactoring [ale-rt]
* Documentation [ale-rt]

0.9.0 (2010/05/05)
------------------

* Initial release [ale-rt]


rt/zps/app.py
# -*- coding: utf-8 -*-
import psutil, sys, getopt

__usage__ = """
USAGE:
  zps
  zps --help
  zps --pid 1111
  zps --port 8080
""".lstrip()

__doc__ = ("""
zps - report a snapshot of the current zope processes.

%s
""" % __usage__).lstrip()

PID_ERROR = "pid value must be an integer number\n\n" + __doc__
REPORT_TEMPLATE=''
def onerror(msg):
    """
    Outputs msg and then exits
    """
    sys.stderr.write(msg)
    return sys.exit()

def checkopt():
    """
    Checking command line options
    """
    flags={'pid': 0, 'port': ''}    
    opt_short=''
    opt_long=['pid=', 'port=']
    try:
        opts, args = getopt.gnu_getopt(sys.argv[1:], opt_short, opt_long)
    except getopt.GetoptError, e:
        onerror("ERROR: %s\n\n%s"%(e, __usage__))
    for opt, optarg in opts:
        if opt == '--pid':
            if not optarg.isdigit():
                onerror(PID_ERROR)
            flags['pid'] = int(optarg)
        if opt == '--port':
            flags['port'] = optarg
    return opts, args, flags

opts, args, flags = checkopt()

class ZProcessReport(dict):
    '''
    The Zope process report class
    '''
    def getcwd(self):
        '''
        Get the current working directory from the process. I found this 
        information missing on a Leopard installation
        '''
        try:
            return self.process.getcwd()
        except AttributeError:
            return "Information not available on this platform"
    
    def getzoncf(self):
        """
        Try to the the zope configuration file
        """
        for x in self.process.cmdline:
            if 'zope.conf' in x:
                return x
        return ''        
    
    def lineparser(self, line, target):
        """
        Parses a line containing target
        """
        if target in line:
            line=line.strip()
            if line.startswith(target):
                return line.replace(target, '').strip()
        return False
    
    def getport(self):
        """
        Tries to get the port of the running zope instance parsing the zope 
        configuration file
        """
        address = False
        pbase = False
        for line in file(self['zconf']):
            if not address:
                address = self.lineparser(line, 'address')
            if not pbase:
                pbase = self.lineparser(line, 'port-base')
            if address and pbase:
                # we already have all the info we need, it is useless to go on
                break
        try:
            return str(int(address)+int(pbase))
        except Exception, e:
            print str(e)
            return ''
    
    def __init__(self, process):
        """
        This reports on a zope process
        """
        self.process = process
        self['cwd'] = self.getcwd()
        self['user'] = process.username
        self['pid'] = process.pid
        # Trying to get from the commandline the zope configuraton file
        self['zconf'] = self.getzconf()
        # Getting the address information from the zope.configuraton file
        self['address'] = self.getport()
        self['memory'] = "%.2f%%" % process.get_memory_percent()
        
    def __str__(self):
        """
        String representation of my process
        """
        return REPORT_TEMPLATE % self
        
class ZProcessFinder(object):
    """
    This object finds out the processes run by zope
    """
    def __init__(self):
        """
        Initialize this object with flags
        """
        
    @property    
    def plist(self):
        """
        This is the list of processes this object is aware of
        """
        plist = []
        for p in psutil.get_process_list():
            pstr = "".join(p.cmdline)
            if 'zope.conf' in pstr:
                pdict = ZProcessReport(p)
                if ((not flags['pid'] or (int(flags['pid']) == pdict['pid'])) 
                    and flags['port'] in pdict['address']):
                    plist.append(pdict)
        return plist

    def __str__(self):
        """
        Returns the output message with info about zope processes
        """
        plist = self.plist
        if not plist:
            return "No running zope instance found"
        return "\n".join(map(str, plist))
